Reland "RegisterCoalescer: Add implicit-def of super register when coalescing SUBREG_TO_REG"

sdesmalen-arm opened the pull request:
This PR aims to reland work done by @arsenm which was previously reverted
due to some tangentially related scheduler issues as discussed on #76416.

This PR cherry-picks the original commit (0e46b49de433), and adds another
patch on top with the following changes:

* The code in `updateRegDefsUses` now updates subranges when
  subreg-liveness-tracking is enabled.

* When adding an implicit-def operand for the super-register,
  the code in `reMaterializeTrivialDef` which tries to remove
  undefined subranges should now take into account that the lanes
  from the super-reg are no longer undefined.

Original patch by @arsenm.

arsenm committed:
Reapply "RegisterCoalescer: Add implicit-def of super register when coalescing SUBREG_TO_REG"

This reverts commit c398fa009a47eb24f88383d5e911e59e70f8db86.

PPC backend was fixed in 2f82662ce901c6666fceb9c6c5e0de216a1c9667

sdesmalen-arm committed:
[RegisterCoalescer] Further changes on top

Changes made compared to 0e46b49de433:

* The code in `updateRegDefsUses` now updates subranges when
  subreg-liveness-tracking is enabled.

* When adding an implicit-def operand for the super-register,
  the code in `reMaterializeTrivialDef` which tries to remove
  undefined subranges should now take into account that the lanes
  from the super-reg are no longer undefined.

* Any operand that defines a register with a sub-reg, must set
  FullDef to false, because it only defines part of the register.

* Ensures that live-ranges are computed for all regunits of a
  physical reg that's added as implicit-def, as needed for subsequent
  passes, see the conversation on:

    https://github.com/llvm/llvm-project/issues/76416

arsenm commented:
Can you add the reported regression test?

arsenm commented on llvm/lib/CodeGen/RegisterCoalescer.cpp:
```diff
@@ -1443,6 +1448,9 @@ bool RegisterCoalescer::reMaterializeTrivialDef(const CoalescerPair &CP,
               (MO.getSubReg() == 0 && MO.getReg() == DstOperand.getReg())) &&
              "unexpected implicit virtual register def");
       ImplicitOps.push_back(MO);
+      if (MO.isDef() && MO.getReg().isVirtual() &&
+          MRI->shouldTrackSubRegLiveness(DstReg))
+        NewMIImplicitOpsMask |= MRI->getMaxLaneMaskForVReg(MO.getReg());
```
Why is this mixing MO.getReg() and DstReg?

sdesmalen-arm committed:
Address review comments

Adds reproducer to show previous regressions are gone.

Also remove code that calculates live ranges for physregs, as
I don't have any tests that cover this case, and the previous
reproducers don't trigger this code. This suggests to me that
the code in #121734 may be sufficient to fix this issue.

arsenm commented on llvm/test/CodeGen/X86/pr76416.ll:
```diff
@@ -0,0 +1,125 @@
+  call void asm sideeffect "", "N{dx},~{dirflag},~{fpsr},~{flags}"(i32 poison) nounwind
+  %2 = load ptr, ptr @load_p, align 8
+  %regs = getelementptr inbounds { [4 x i8] }, ptr %2, i32 0, i32 0
+  %3 = load i32, ptr %i, align 4
```
Use named values in tests

arsenm commented on llvm/test/CodeGen/X86/pr76416.ll:
```diff
@@ -0,0 +1,125 @@
+; Related reproducer as reported on https://github.com/llvm/llvm-project/commit/0e46b49de43349f8cbb2a7d4c6badef6d16e31ae#commitcomment-136147998
+;
+
+define void @f(i1 %cmp.not.i.i.i) {
```
I think this one is the same as  llvm/test/CodeGen/X86/coalescer-add-implicit-def-subreg-to-reg-regression.ll

sdesmalen-arm committed:
Name values in test, and remove duplicate test.
